### PR TITLE
feat: add app.getApplicationInfoForProtocol API 

### PR DESCRIPTION
**Pull request by georgexu99**
#### Description of Change

This api returns the name, display icon and path of the application registered as the default handler of a specific protocol (eg. the protocol 'https://example.com' would return 'Chrome', Chrome's icon and its installation path (or whichever your default browser happens to be). 

It's an extension of the "[getApplicationNameForProtocol](https://github.com/electron/electron/pull/20399)" api to allow for more information on the application to be returned 

#### Checklist

- [x] PR description included and stakeholders cc'd
- [x] `npm test` passes
- [x] tests are added (https://github.com/electron/electron/blob/master/docs/development/testing.md)
- [x] relevant documentation is added
- [x] PR title follows semantic [commit guidelines](https://github.com/electron/electron/blob/master/docs/development/pull-requests.md#commit-message-guidelines)
- [x] [PR release notes](https://github.com/electron/clerk/blob/master/README.md) describe the change in a way relevant to app developers, and are [capitalized, punctuated, and past tense](https://github.com/electron/clerk/blob/master/README.md#examples).
- [x] This is **NOT A BREAKING CHANGE**. Breaking changes may not be merged to master until 11-x-y is branched.

#### Release Notes

Notes: Added `app.getApplicationInfoForProtocol()` API that returns detailed information about the app that handles a certain protocol

### CODE REVIEW

File: spec-main/api-app-spec.ts [hunk 1]
```diff
@@ -973,6 +973,31 @@ describe('app module', () => {
     });
   });
 
+  describe('getApplicationInfoForProtocol()', () => {
```
Comment (codebytere) [inline spec-main/api-app-spec.ts:976]: We can use `ifdescribe(process.platform !== 'linux')` here and remove the skip conditionals from individual tests

File: shell/browser/browser.h [hunk 2]
```diff
@@ -98,6 +100,12 @@ class Browser : public WindowListObserver {
 
   base::string16 GetApplicationNameForProtocol(const GURL& url);
 
+#if !defined(OS_LINUX)
+  // get the name, icon and path for an application
+  v8::Local<v8::Promise> GetApplicationInfoForProtocol(const GURL& url,
```
Comment (codebytere) [inline shell/browser/browser.h:105]: By convention, `v8::Isolate* isolate` (&`gin::Arguments* args`) should go first in a parameter list

Comment (georgexu99) [reply shell/browser/browser.h]: thank you!

File: spec-main/api-app-spec.ts (hunk 1 above)
Comment (georgexu99) [reply spec-main/api-app-spec.ts:976]: thank you

File: docs/api/app.md [hunk 3]
```diff
@@ -805,6 +805,20 @@ Returns `String` - Name of the application handling the protocol, or an empty
 This method returns the application name of the default handler for the protocol
 (aka URI scheme) of a URL.
 
+### `app.getApplicationInfoForProtocol(url)`
```
Comment (MarshallOfSound) [inline docs/api/app.md:808]: This should be commented as `_macOS_ _Windows_` as it isn't supported on Linux

File: shell/browser/browser_mac.mm [hunk 4]
```diff
@@ -147,20 +148,68 @@
   return result == NSOrderedSame;
 }
 
-base::string16 Browser::GetApplicationNameForProtocol(const GURL& url) {
+gfx::Image GetApplicationIconForProtocol(NSString* _Nonnull app_path) {
```
Comment (MarshallOfSound) [inline shell/browser/browser_mac.mm:151]: This should be in the anonymous namespace

File: shell/browser/browser_mac.mm [hunk 5]
```diff
@@ -147,20 +148,68 @@
...
+  NSImage* image = [[NSWorkspace sharedWorkspace] iconForFile:app_path];
+  gfx::Image icon(image);
+  return icon;
+}
+
+base::string16 GetAppDisplayNameForProtocol(NSString* app_path) {
```
Comment (MarshallOfSound) [inline shell/browser/browser_mac.mm:157]: Same here, anonymous namespace

File: shell/browser/browser_mac.mm [hunk 6]
```diff
@@ -147,20 +148,68 @@
...
+  NSString* app_display_name =
+      [[NSFileManager defaultManager] displayNameAtPath:app_path];
+  return base::SysNSStringToUTF16(app_display_name);
+}
+
+NSString* GetAppPathForProtocol(const GURL& url) {
```
Comment (MarshallOfSound) [inline shell/browser/browser_mac.mm:163]: Again, anonymous namespace

File: shell/browser/browser_win.cc [hunk 7]
```diff
@@ -105,6 +111,62 @@ base::string16 GetAppForProtocolUsingAssocQuery(const GURL& url) {
...
+    promise.RejectWithErrorMessage("Failed to get file icon.");
+  }
+}
+
+base::string16 GetAppDisplayNameForProtocol(const GURL& url) {
+  return GetAppInfoHelperForProtocol(ASSOCSTR_FRIENDLYAPPNAME, url);
+}
+
+base::string16 GetAppPathForProtocol(const GURL& url) {
+  return GetAppInfoHelperForProtocol(ASSOCSTR_FRIENDLYAPPNAME, url);
```
Comment (MarshallOfSound) [inline shell/browser/browser_win.cc:163]: `ASSOCSTR_FRIENDLYAPPNAME` ? Copy-paste issue?

File: shell/browser/browser_win.cc [hunk 8]
```diff
@@ -105,6 +111,62 @@ base::string16 GetAppForProtocolUsingAssocQuery(const GURL& url) {
...
+}
+
+base::string16 GetApplicationIconForProtocol(const GURL& url) {
+  return GetAppInfoHelperForProtocol(ASSOCSTR_FRIENDLYAPPNAME, url);
```
Comment (MarshallOfSound) [inline shell/browser/browser_win.cc:167]: `ASSOCSTR_FRIENDLYAPPNAME` ? Copy-paste issue?

File: shell/browser/browser_win.cc [hunk 9]
```diff
@@ -105,6 +111,62 @@ base::string16 GetAppForProtocolUsingAssocQuery(const GURL& url) {
...
+    dict.Set("appPath", app_path);
+    dict.Set("displayName", app_display_name);
+    dict.Set("icon", icon);
+    promise.Resolve(dict);
+  } else {
+    promise.RejectWithErrorMessage("Failed to get file icon.");
+  }
+}
+
+base::string16 GetAppDisplayNameForProtocol(const GURL& url) {
```
Comment (MarshallOfSound) [inline shell/browser/browser_win.cc:158]: You should remove `GetAppForProtocolUsingAssocQuery` from this file and replace it's usages with this new `GetAppDisplayNameForProtocol`.  It's the same code / result right?

File: shell/browser/browser_win.cc [hunk 10]
```diff
@@ -169,6 +231,91 @@ void Browser::Focus(gin_helper::Arguments* args) {
...
 }
 
+void GetFileIcon(const base::FilePath& path,
+                 v8::Isolate* isolate,
+                 base::CancelableTaskTracker* cancelable_task_tracker_,
+                 const base::FilePath& app_path,
+                 base::string16 app_display_name,
+                 gin_helper::Promise<gin_helper::Dictionary> promise) {
+  base::FilePath normalized_path = path.NormalizePathSeparators();
+  gin_helper::Dictionary dict = gin::Dictionary::CreateEmpty(isolate);
```
Comment (MarshallOfSound) [inline shell/browser/browser_win.cc:241]: We only need to construct this dict if `LookupIconFromFilepath` returns.  We can move this allocation to inside the `if (icon) {` block

File: shell/browser/browser_win.cc [hunk 11]
```diff
@@ -169,6 +231,91 @@ void Browser::Focus(gin_helper::Arguments* args) {
   EnumWindows(&WindowsEnumerationHandler, reinterpret_cast<LPARAM>(&pid));
 }
 
+void GetFileIcon(const base::FilePath& path,
+                 v8::Isolate* isolate,
+                 base::CancelableTaskTracker* cancelable_task_tracker_,
+                 const base::FilePath& app_path,
+                 base::string16 app_display_name,
```
Comment (MarshallOfSound) [inline shell/browser/browser_win.cc:238]: This can be a const reference

File: shell/browser/browser_win.cc [hunk 12]
```diff
@@ -105,6 +111,62 @@ base::string16 GetAppForProtocolUsingAssocQuery(const GURL& url) {
...
+                       out_buffer, &buffer_size);
+  if (FAILED(hr)) {
+    DLOG(WARNING) << "AssocQueryString failed!";
+    return base::string16();
+  }
+  return base::string16(out_buffer);
+}
+
+void OnIconDataAvailable(base::FilePath app_path,
+                         base::string16 app_display_name,
```
Comment (MarshallOfSound) [inline shell/browser/browser_win.cc:141]: Can this be a const reference?

File: shell/browser/browser_win.cc [hunk 13]
```diff
@@ -105,6 +111,62 @@ base::string16 GetAppForProtocolUsingAssocQuery(const GURL& url) {
...
+      AssocQueryString(ASSOCF_IS_PROTOCOL, assoc_str, url_scheme.c_str(), NULL,
+                       out_buffer, &buffer_size);
+  if (FAILED(hr)) {
+    DLOG(WARNING) << "AssocQueryString failed!";
+    return base::string16();
+  }
+  return base::string16(out_buffer);
+}
+
+void OnIconDataAvailable(base::FilePath app_path,
```
Comment (MarshallOfSound) [inline shell/browser/browser_win.cc:140]: Can this be a const reference?

File: shell/browser/browser_win.cc [hunk 14]
```diff
@@ -169,6 +231,91 @@ void Browser::Focus(gin_helper::Arguments* args) {
...
+  }
+}
+
+void GetApplicationInfoForProtocolUsingRegistry(
+    v8::Isolate* isolate,
+    const GURL& url,
+    gin_helper::Promise<gin_helper::Dictionary> promise,
+    base::CancelableTaskTracker* cancelable_task_tracker_) {
+  base::FilePath app_path;
+  const base::string16 app_display_name = GetAppForProtocolUsingRegistry(url);
```
Comment (MarshallOfSound) [inline shell/browser/browser_win.cc:266]: This should be moved to after we check if the `url_scheme` is valid

File: shell/browser/browser_win.cc [hunk 15]
```diff
@@ -105,6 +111,62 @@ base::string16 GetAppForProtocolUsingAssocQuery(const GURL& url) {
...
+
+base::string16 GetAppDisplayNameForProtocol(const GURL& url) {
+  return GetAppInfoHelperForProtocol(ASSOCSTR_FRIENDLYAPPNAME, url);
+}
+
+base::string16 GetAppPathForProtocol(const GURL& url) {
+  return GetAppInfoHelperForProtocol(ASSOCSTR_FRIENDLYAPPNAME, url);
+}
+
+base::string16 GetApplicationIconForProtocol(const GURL& url) {
```
Comment (MarshallOfSound) [inline shell/browser/browser_win.cc:166]:
```suggestion
base::string16 GetApplicationIconPathForProtocol(const GURL& url) {
```

File: shell/browser/browser_win.cc [hunk 16]
```diff
@@ -169,6 +231,91 @@ void Browser::Focus(gin_helper::Arguments* args) {
...
+    v8::Isolate* isolate,
+    const GURL& url,
+    gin_helper::Promise<gin_helper::Dictionary> promise,
+    base::CancelableTaskTracker* cancelable_task_tracker_) {
+  base::string16 app_path = GetAppPathForProtocol(url);
+
+  if (app_path.empty()) {
+    promise.RejectWithErrorMessage(
+        "Unable to retrieve installation path to app");
+  }
```
Comment (MarshallOfSound) [inline shell/browser/browser_win.cc:304]: Need to `return` after this rejection

File: shell/browser/browser_win.cc [hunk 17]
```diff
@@ -169,6 +231,91 @@ void Browser::Focus(gin_helper::Arguments* args) {
...
+
+  base::string16 app_display_name = GetAppDisplayNameForProtocol(url);
+
+  if (app_display_name.empty()) {
+    promise.RejectWithErrorMessage("Unable to retrieve display name of app");
+  }
```
Comment (MarshallOfSound) [inline shell/browser/browser_win.cc:310]: Need to `return` after this rejection

File: shell/browser/browser_win.cc [hunk 18]
```diff
@@ -81,30 +86,58 @@ bool IsValidCustomProtocol(const base::string16& scheme) {
...
+                       out_buffer, &buffer_size);
   if (FAILED(hr)) {
     DLOG(WARNING) << "AssocQueryString failed!";
     return base::string16();
   }
   return base::string16(out_buffer);
 }
 
+void OnIconDataAvailable(const base::FilePath app_path,
+                         const base::string16 app_display_name,
```
Comment (codebytere) [inline shell/browser/browser_win.cc:116]: `app_path` and `app_display_name` can both be const refs

**Comment by jkleinsc** [comment]
@electron/wg-api reviewed and approved this at our June 29, 2020 meeting

**Pull request merged**